File: random-seating/random-seating.cpp
#include <iostream>
#include <fstream>
#include <cstdio>

#include <algorithm>
#include <time.h>

#include <vector>
#include <map>

#include <stdexcept>

void
get_good_shuf(std::vector<size_t>& arr, const size_t& len)
{
	arr.resize(len);
	int i, n, j;
	for (i = 0, n = arr.size(); i < n; i++)
		arr[i] = i;

	for (i = arr.size()-1, n = arr.size(); i >= 0; i--)
		std::swap(arr[i], arr[ rand() % n ]);

	if (len <= 3) return;

	bool changes = true;
	while (changes) {
		changes = false;
		for (i = 1, n = arr.size(); i < n; i++) {
			if ( std::abs(arr[i-1] - arr[i]) <= 1 ) {

				do {
					j = rand() % n;
				} while (j == i-1 || j == i);

				std::swap(arr[i], arr[j]);
				changes = true;
			}
		}
	}
}

int main(int argc, char **argv) {
	srand(time(NULL));

	if (argc != 2) {
		std::cout << "Usage: ./random-seating [seating file]\n";
		exit(1);
	}

	std::fstream fs;
	try {
		fs = std::fstream(argv[1], std::fstream::in);
	} catch (...) {
		std::cout << "Could not open " << std::string(argv[1]) << std::endl;
		exit(2);
	}

	std::vector<std::string> students_orig;
	std::string tmp;
	while(std::getline(fs, tmp)) {
		students_orig.push_back(tmp);
	}

	/* close file stream*/
	fs.close();

	std::vector<size_t> arr;
	get_good_shuf(arr, students_orig.size());

	for (size_t i = 0; i < arr.size(); i++)
		std::cout << i+2 << "|\t" << students_orig[arr[i]] << std::endl;
}

File: random-seating/random-seating.rb
=begin

random-seating.rb

A port to Ruby of randomSeating.py.

@author Charles Bailey
@version 0.2 9/26/16

=end

def get_good_shuf(len)
	arr = []
	len.times{|i| arr << i}
	arr.shuffle!

	return arr if len <= 3

	changes = true
	while changes
		changes = false

		for i in 1...arr.length
			if (arr[i-1] - arr[i]).abs <= 1
				j = rand(0...arr.length)
				while (j == i-1 || j == i)
					j = rand(0...arr.length)
				end

				arr[i-1],arr[j] = arr[j],arr[i-1]
				changes = true
			end
		end

	end
	return arr
end


# main routine:
unless ARGV.length == 1
	puts "[!] Usage: ./random-seating.rb [seating chart file]"
	exit(1)
end

begin
	students_original = File.readlines(ARGV[0]).map{|line| line.strip}
rescue
	puts "[!] Could not open file #{ARGV[0]}"
	exit(2)
end

arr = get_good_shuf(students_original.length)
arr.each_with_index do |v,i|
	puts "#{format("%2i", i+2)}|\t#{students_original[v]}"
end
